src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,218 @@
+
+std::optional<request_ptr> requests::last_request() const {
+    if (!_inflight_requests.empty()) {
+        return _inflight_requests.back();
+    } else if (!_finished_requests.empty()) {
+        return _finished_requests.back();
+    }
+    return std::nullopt;
+}
+
+requests::request_result_t
+requests::new_request(seq_t first, seq_t last, model::term_id current) {
+    vlog(
+      clusterlog.trace,
+      "new request attempt, first: {}, last: {}, term: {}, producer: {}",
+      first,
+      last,
+      current,
+      *this);
+    // gc any inflight requests from old terms
+    while (!_inflight_requests.empty()
+           && _inflight_requests.front()->_term < current
+           && _inflight_requests.front()->in_progress()) {
+        // Here we know for sure the term change, these in flight requests
+        // are going to fail anyway, mark them so.
+        _inflight_requests.front()->set_value(errc::not_leader);
+        _inflight_requests.pop_front();
+    }
+
+    // check if an existing request matches
+    auto match_it = std::find_if(
+      _finished_requests.begin(),
+      _finished_requests.end(),
+      [first, last](const auto& request) {
+          return request->_first_sequence == first
+                 && request->_last_sequence == last;
+      });
+
+    if (match_it != _finished_requests.end()) {
+        return match_it->get()->result.get_shared_future();
+    }
+
+    match_it = std::find_if(
+      _inflight_requests.begin(),
+      _inflight_requests.end(),
+      [first, last, current](const auto& request) {
+          return request->_first_sequence == first
+                 && request->_last_sequence == last
+                 && request->_term == current;
+      });
+
+    if (match_it != _inflight_requests.end()) {
+        return match_it->get()->result.get_shared_future();
+    }
+
+    // check if the incoming request forms a sequence.
+    auto req = last_request();
+    if (
+      // there is a prev request, ensure sequence invariant is maintained.
+      (req && req.value()->_last_sequence + 1 != first)
+      // there is no prev request, incoming sequence number must be 0.
+      || (!req && first != 0)) {
+        return errc::sequence_out_of_order;
+    }
+
+    // All invariants satisfied, enqueue the request.
+    _inflight_requests.emplace_back(
+      ss::make_lw_shared<request>(first, last, current, result_promise_t{}));
+
+    return _inflight_requests.back();
+}
+
+bool requests::stm_apply(
+  const model::batch_identity& bid, kafka::offset offset) {
+    bool is_follower = false;
+    auto first = bid.first_seq;
+    auto last = bid.last_seq;
+    if (!_inflight_requests.empty()) {
+        auto front = _inflight_requests.front();
+        if (front->_first_sequence == first && front->_last_sequence == last) {
+            // Promote the request from in_flight -> finished.
+            _inflight_requests.pop_front();
+        }
+    } else {
+        is_follower = true;
```

Comment:
it isn't obvious what `is_follower` mean. a node may just become a leader and while it replays writes of previous leader its _inflight_requests is empty